README.md
Riak-Light
==========

Fast and lightweight Perl client for Riak [![CPAN version](https://badge.fury.io/pl/Riak-Light.png)](http://badge.fury.io/pl/Riak-Light)

```perl
    # create a new instance - using pbc only
    my $client = Riak::Light->new(
      host => '127.0.0.1',
      port => 8087
    );
    
    $client->is_alive() or die "ops, riak is not alive";

    # store hashref into bucket 'foo', key 'bar'
    # will serializer as 'application/json'
    $client->put( foo => bar => { baz => 1024 });
    
    # store text into bucket 'foo', key 'bar' 
    $client->put( foo => baz => "sometext", 'text/plain');
    $client->put_raw( foo => baz => "sometext");  # does not encode !

    # fetch hashref from bucket 'foo', key 'bar'
    my $hash = $client->get( foo => 'bar');
    my $text = $client->get_raw( foo => 'baz');   # does not decode !

    # delete hashref from bucket 'foo', key 'bar'
    $client->del(foo => 'bar');
    
    # check if exists (like get but using less bytes in the response)
    $client->exists(foo => 'baz') or warn "ops, foo => bar does not exist";
    
    # list keys in stream (callback only)
    $client->get_keys(foo => sub{
       my $key = $_[0];
       
       # you should use another client inside this callback!
       $another_client->del(foo => $key);
    });
```

Install
=======

It is available on CPAN

    cpan Riak::Light
    
https://metacpan.org/release/Riak-Light/

Test Coverage
=============

    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    File                           stmt   bran   cond    sub    pod   time  total
    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    blib/lib/Riak/Light.pm        100.0  100.0  100.0  100.0   80.0   48.4   99.5
    ...b/Riak/Light/Connector.pm  100.0  100.0    n/a  100.0    0.0   32.4   96.6
    .../lib/Riak/Light/Driver.pm  100.0  100.0    n/a  100.0    0.0    4.0   93.3
    blib/lib/Riak/Light/PBC.pm    100.0    n/a    n/a  100.0    n/a   10.1  100.0
    ...lib/Riak/Light/Timeout.pm  100.0    n/a    n/a  100.0    n/a    0.7  100.0
    ...ak/Light/Timeout/Alarm.pm  100.0  100.0    n/a  100.0    0.0    0.7   94.3
    ...k/Light/Timeout/Select.pm  100.0    n/a    n/a  100.0    0.0    0.4   90.2
    ...t/Timeout/SelectOnRead.pm  100.0    n/a    n/a  100.0    0.0    0.2   91.3
    ...ght/Timeout/SetSockOpt.pm  100.0  100.0    n/a  100.0    0.0    0.5   96.8
    .../Light/Timeout/TimeOut.pm  100.0  100.0    n/a  100.0    0.0    0.9   94.9
    blib/lib/Riak/Light/Util.pm   100.0    n/a    n/a  100.0    0.0    1.7   87.5
    Total                         100.0  100.0  100.0  100.0   15.4  100.0   96.1
    ---------------------------- ------ ------ ------ ------ ------ ------ ------


Simple Benchmark
================

Using Perl 5.12.2 under MacOSX 10.8.3 / 4GB Ram / 2.4 GHz Intel Core 2 Duo and Riak 1.3.0 (localhost)

Only GET (`benchmark/compare_all_only_get.pl`)
 
                             Rate Data::Riak (REST) Net::Riak (REST) Riak::Tiny (REST) Data::Riak::Fast (REST) Net::Riak (PBC) Riak::Light (PBC)
    Data::Riak (REST)        318/s                --             -33%              -42%                    -46%            -69%              -92%
    Net::Riak (REST)         478/s               50%               --              -12%                    -20%            -54%              -87%
    Riak::Tiny (REST)        544/s               71%              14%                --                     -8%            -47%              -86%
    Data::Riak::Fast (REST)  594/s               87%              24%                9%                      --            -42%              -84%
    Net::Riak (PBC)         1031/s              224%             116%               89%                     73%              --              -73%
    Riak::Light (PBC)       3774/s             1086%             690%              593%                    535%            266%                --               --
 
Only PUT (`benchmark/compare_all_only_put.pl`)

                              Rate Net::Riak (REST) Data::Riak (REST) Riak::Tiny (REST) Net::Riak (PBC) Data::Riak::Fast (REST) Riak::Light (PBC)
    Net::Riak (REST)         418/s               --              -16%              -22%            -55%                    -58%              -88%
    Data::Riak (REST)        499/s              19%                --               -7%            -46%                    -50%              -86%
    Riak::Tiny (REST)        534/s              28%                7%                --            -42%                    -46%              -85%
    Net::Riak (PBC)          924/s             121%               85%               73%              --                     -6%              -74%
    Data::Riak::Fast (REST)  988/s             137%               98%               85%              7%                      --              -73%
    Riak::Light (PBC)       3604/s             763%              623%              575%            290%                    265%                --

Timeout Providers (`benchmark/compare_timeout_providers.pl`)

                    Rate Riak::Light 6 Riak::Light 3 Riak::Light 2 Riak::Light 4 Riak::Light 5 Riak::Light 1
    Riak::Light 6 2158/s            --          -16%          -21%          -21%          -32%          -38%
    Riak::Light 3 2564/s           19%            --           -6%           -6%          -20%          -26%
    Riak::Light 2 2727/s           26%            6%            --            0%          -15%          -22%
    Riak::Light 4 2727/s           26%            6%            0%            --          -15%          -22%
    Riak::Light 5 3191/s           48%           24%           17%           17%            --           -9%
    Riak::Light 1 3488/s           62%           36%           28%           28%            9%            --

    where

    1 - undef / no IO timeout
    2 - using Time::HiRes (alarm)
    3 - using IO::Select (DEFAULT)
    4 - using IO::Select only in read operations
    5 - using setsockopt
    6 - using Time::Out

Features
========

* be PBC only (ok)
* supports timeout (ok)
* use Moo (ok)
* doesn't create an object per key (ok)
* support an option to not die, but return undef (deprecated, will be removed to the next version)
* be optimized for speed. (ok)
* try to get 100% coverage. (ok)
* benchmark with Data::Riak, Net::Riak REST, etc... (ok)
* documentation (ok)
* support raw data (ok)
* support list keys (ok)
* debug/trace mode (to do)
* refactor to use Moo::Roles instead other objects (to do)
* extract timeout provider to an external project (to do)


Changes
0.11 2014-09-23 12:00:00 GMT
  * send extra parameter r,w, dw, pr, pw, rw in all requests, if necessary (Tiago Peczenyj)

0.10 2014-08-31 19:00:00 GMT
  * fork safe capability (Tiago Peczenyj)

0.09911 2014-03-03 14:42:00 GMT
   * fix get_index_value, add get_all_index_values (Tiago Peczenyj) 

0.09901 2014-02-28 13:00:00 GMT
   * supports set/get client id (Tiago Peczenyj)

0.099 2014-02-28 12:00:00 GMT
   * first steps to support vclock (Tiago Peczenyj)

0.084 2013-12-13 14:00:00 GMT
   * improve map indexes (Tiago Peczenyj)

0.083 2013-12-13 12:00:00 GMT
   * add get_all_indexes method (Tiago Peczenyj)

0.082 2013-12-06 14:00:00 GMT
   * map reduce accept string or hashref query (Tiago Peczenyj)

0.081 2013-12-06 13:00:00 GMT
   * fix documentatiom (Tiago Peczenyj)

0.080 2013-12-06 12:00:00 GMT
   * Support to map_reduce operations (Tiago Peczenyj)

0.072 2013-11-15 17:26:00 GMT 
   * TCP_NODELAY should be configurable (Tiago Peczenyj)

0.071 2013-11-15 16:55:00 GMT 
   * Add TCP_NODELAY 1 to socket (Tiago Peczenyj)

0.070 2013-11-15 15:43:00 GMT 
   * Support pagination results (Tiago Peczenyj)

0.061 2013-09-10 15:39:00 GMT 
   * Support multi-values in secondary indexes (Damien Krotkine)

0.060 2013-07-25 15:22:00 GMT 
   * Support secondary indexes (2i)

0.059 2013-06-06 16:27:08 GMT 
   * Refactor util class. (Tiago Peczenyj)

0.058 2013-06-06 15:27:08 GMT 
   * Small refactor, fix small things, fix solaris support. (Tiago Peczenyj)

0.057 2013-05-30 15:27:08 GMT 
   * Cleanups, speedups, refactoring (Damien Krotkine)

0.056 2013-05-17 08:18:08 GMT 
   * bugfix clean $@ before all methods [issue #9] (Tiago Peczenyj)

0.055 2013-05-16 22:18:08 GMT 
   * remove namespace::autoclean [issue #8] (Tiago Peczenyj)
   * change minimum perl to 5.008

0.054 2013-05-16 11:31:08 GMT 
   * Remove extra test (Tiago Peczenyj) 

0.053 2013-05-16 09:31:08 GMT 
   * Remove extra method (debug / experimental) (Tiago Peczenyj) 

0.052 2013-05-16 07:31:08 GMT 
   * Fix owner

0.051 2013-05-15 18:31:08 GMT 
   * Fix test

0.05 2013-05-15 17:31:08 GMT 
   * Fix bug in EOF
   * add exists method, to test if there is some bucket => key

0.04 2013-05-14 21:21:27 GMT 
   * Fix bug in SetSockOpt (Tiago Peczenyj)
   * Add warnings in Alarm and TimeOut providers (not safe)
   * Add OS type / version check for some providers (Win32, NetBSD)
   * Improve test coverage
   * IO::Select is the default timeout provider
   * Fix tests and error strings (will work for other OS)
   * Minimum perl 5.12
   * Small improvements 

0.03 2013-05-14 11:21:27 GMT 
   * Improve setsockopt (Tiago Peczenyj)
   
0.02 2013-05-13 13:16:08 GMT 
   * Fix documentatio (Tiago Peczenyj)

0.01 2013-05-13 10:16:08 GMT 
   * First version (Tiago Peczenyj)


lib/Riak/Light/PBC.pm
## no critic (RequireUseStrict, RequireUseWarnings)
package Riak::Light::PBC;
## use critic
# ABSTRACT: ProtocolBuffers helper to access the Riak API
##
## This file was generated by Google::ProtocolBuffers (0.08)
## on Fri Feb 28 12:59:39 2014
##      
use strict;
use warnings;
use Google::ProtocolBuffers;
{       
    unless (RpbIndexReq::IndexQueryType->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_enum(
            'RpbIndexReq::IndexQueryType',
            [
               ['eq', 0],
               ['range', 1],

            ]
        );
    }
    
    unless (RpbSetClientIdReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbSetClientIdReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'client_id', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbPutReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbPutReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    'RpbContent', 
                    'content', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'w', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'dw', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'return_body', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pw', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'if_not_modified', 9, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'if_none_match', 10, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'return_head', 11, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'r', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pr', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'basic_quorum', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'notfound_ok', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'if_modified', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'head', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'deletedvclock', 9, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbLink->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbLink',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'tag', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbContent', 
                    'content', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'unchanged', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbPair->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbPair',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'value', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbPutResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbPutResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbContent', 
                    'contents', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbDelReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbDelReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'rw', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'r', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'w', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pr', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pw', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'dw', 9, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbMapRedReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbMapRedReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'request', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'content_type', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbIndexReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbIndexReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'index', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    'RpbIndexReq::IndexQueryType', 
                    'qtype', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'range_min', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'range_max', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'return_terms', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'stream', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'max_results', 9, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'continuation', 10, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetClientIdResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetClientIdResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'client_id', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbMapRedResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbMapRedResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'phase', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'response', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'done', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbIndexResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbIndexResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'keys', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbPair', 
                    'results', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'continuation', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'done', 4, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbErrorResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbErrorResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'errmsg', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'errcode', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbListKeysReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbListKeysReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbListKeysResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbListKeysResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'keys', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'done', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbContent->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbContent',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'value', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'content_type', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'charset', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'content_encoding', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vtag', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbLink', 
                    'links', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'last_mod', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'last_mod_usecs', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbPair', 
                    'usermeta', 9, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbPair', 
                    'indexes', 10, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'deleted', 11, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

}
1;
